### PR TITLE
Unify HuggingFace model loading and more

### PR DESCRIPTION
**Pull request by sdascoli**
## Why this PR?

- Model build, device and dtype handling is extremely messy and ad-hoc in our HuggingFace extractors, and different across modalities.  
- For some HF models, AutoModel/AutoProcessor do not work, you need to load a specific model class such as VivitModel. Currently this behaviour is hardcoded, meaning we only support a fraction of HF models
- We currently do not support giving arguments such as trust_remote_code to the HF model build, which forced people such as @shouhamdi to create custom subclasses for each edge case
- device="accelerate" was an alias for {"device_map": "auto" and "dtype=float16"}, which is quite arbitrary and caused bugs in HuggingFaceImage (see https://github.com/facebookresearch/neuroai/pull/88)

## Summary

Unifies HuggingFace model loading across text, audio, image, and video extractors by moving shared model construction options into a common `hf_config` and centralizing model instantiation in `HuggingFaceMixin`. This gives us full flexibility on model instantiation

## What Changed

- Added `HuggingFaceMixin.load_model()` and shared model/processor caches so text, audio, image, and video extractors use the same HuggingFace loading path.
- Restored explicit `device` and `dtype` fields on `HuggingFaceMixin`: `device="auto"` resolves to CUDA when available otherwise CPU, while `device="accelerate"` is the opt-in path that forwards `device_map="auto"` to HuggingFace.
- Added `HuggingFaceConfig` for HuggingFace class selection and escape-hatch kwargs: `model_cls_name`, `processor_cls_name`, `model_kwargs`, and `processor_kwargs`.
- Added public `HuggingFaceConfig.HF_CLASS_DEFAULTS` so known model-name patterns can resolve non-standard model/processor classes without modality-specific `if model_name` branches.
- Treat audio feature extractors and text tokenizers as shared HuggingFace processors, so modality-specific processor kwargs and caching follow the same path.
- Simplified image and video extractors by removing wrapper classes, merging `BaseImage` into `HuggingFaceImage`, and removing `_HFVideoModel`, `layer_type`, xclip-specific handling, and model-name-based frame defaults.
- Made native `HuggingFaceVideo.num_frames` required and warn when it differs from `model.config.num_frames` when available.
- Kept HuggingFace repo existence validation and added validation errors for misplaced `device_map`/`torch_dtype` entries in `hf_config.model_kwargs`; users should use `device`/`dtype` instead.

### CODE REVIEW

File: neuralset-repo/neuralset/events/transforms/test_transforms.py [hunk 1]
```diff
@@ -1065,7 +1065,10 @@ def test_similarity_splitter(image_fpaths: list[Path]) -> None:
         }
     )
 
-    feat = ns.extractors.image.HuggingFaceImage(device="cpu", layers=1.0)
+    feat = ns.extractors.image.HuggingFaceImage(
+        hf_config={"device_map": "cpu"},  # type: ignore[arg-type]
```
Comment (kingjr) [inline neuralset-repo/neuralset/events/transforms/test_transforms.py]: not a huge fan of the variable names

File: neuralset-repo/neuralset/extractors/audio.py [hunk 2]
```diff
@@ -379,24 +379,18 @@ def feature_extractor(self) -> nn.Module:
             self._feature_extractor = self._get_feature_extractor(self.model_name)
         return self._feature_extractor
 
-    @property
-    def model(self) -> nn.Module:
-        if not hasattr(self, "_model"):
-            self._model = self._get_sound_model(self.model_name)
-        return self._model
+    def _load_cached_model(self) -> torch.nn.Module:
+        return self._get_sound_model(self.model_name)
```
Comment (kingjr) [inline neuralset-repo/neuralset/extractors/audio.py]: do we need these nested methods?

File: neuralset-repo/neuralset/extractors/audio.py [hunk 3]
```diff
@@ -476,14 +470,9 @@ class Wav2VecBert(HuggingFaceAudio):
...
-
-    def _get_sound_model(self, model_name: str) -> torch.nn.Module:
-        from transformers import Wav2Vec2BertModel
-
-        _model = Wav2Vec2BertModel.from_pretrained(model_name)
-        _model.to(self.device)
-        _model.eval()
-        return _model
+    hf_config: HuggingFaceConfig = HuggingFaceConfig(
+        model_cls_name="Wav2Vec2BertModel",
```
Comment (kingjr) [inline neuralset-repo/neuralset/extractors/audio.py:468]: why not model_name like the rest of the api?

File: neuralset-repo/neuralset/extractors/audio.py [hunk 4]
```diff
@@ -502,14 +491,12 @@ class SeamlessM4T(HuggingFaceAudio):
     """
 
     model_name: str = "facebook/hf-seamless-m4t-medium"
```
Comment (kingjr) [inline neuralset-repo/neuralset/extractors/audio.py:487]: Is there an argument to have this dissociation between model_name and hf_config? model_name could have been part of the hf_config OR hf_config at the root. This dissociation feels a bit clunky.

Maybe it's just me / a varname issue, e.g. model_infra?

File: neuralset-repo/neuralset/extractors/base.py [hunk 5]
```diff
@@ -479,63 +592,75 @@ class HuggingFaceMixin(base.BaseModel):
...
+        else:
+            config = AutoConfig.from_pretrained(
+                self.model_name,
+                **hf_config.config_kwargs,
+            )
+            model = Model.from_config(config)
+            if isinstance(torch_dtype, torch.dtype):
+                model.to(dtype=torch_dtype)
+            model.to(self.device)
+        model.eval()
```
Comment (kingjr) [inline neuralset-repo/neuralset/extractors/base.py:711]: naive question: do we usually set eval by default?

File: neuralset-repo/neuralset/extractors/base.py [hunk 6]
```diff
@@ -438,6 +437,120 @@ def _skip_new_event_types(key, val, prev):
...
+            "float32",
+            "float64",
+            "bfloat16",
+        ]
+        | None
+    ) = None
+    attn_implementation: str | None = None
+    revision: str | None = None
+    trust_remote_code: bool = False
+    model_cls_name: str | None = None
```
Comment (kingjr) [inline neuralset-repo/neuralset/extractors/base.py]: when can this be None? SHouldn't we set to AutoModel?

File: neuralset-repo/neuralset/extractors/base.py [hunk 7]
```diff
@@ -438,6 +437,120 @@ def _skip_new_event_types(key, val, prev):
...
+    processor_cls_name: str | None = None
+
+    @classmethod
+    def _exclude_from_cls_uid(cls) -> list[str]:
+        return ["device_map", "attn_implementation", "trust_remote_code"]
```
Comment (kingjr) [inline neuralset-repo/neuralset/extractors/base.py]: are we sure attn_implementation doesn't affect the results?

(same question for device actually, I suspect there are edge cases but we'd like the device to not matter)

File: neuralset-repo/neuralset/extractors/base.py [hunk 8]
```diff
@@ -438,6 +437,120 @@ def _skip_new_event_types(key, val, prev):
...
+
+    def processor_cls(self) -> tp.Any:
+        return self._transformers_cls(
+            self.processor_cls_name,
+            "AutoProcessor",
+            "processor",
+        )
+
+    @property
+    def config_kwargs(self) -> dict[str, tp.Any]:
```
Comment (kingjr) [inline neuralset-repo/neuralset/extractors/base.py]: .config? .to_dict()?

File: neuralset-repo/neuralset/extractors/base.py [hunk 9]
```diff
@@ -438,6 +437,120 @@ def _skip_new_event_types(key, val, prev):
...
+        return {
+            key: val
+            for key, val in {
+                "revision": self.revision,
+                "trust_remote_code": self.trust_remote_code or None,
+            }.items()
+            if val is not None
+        }
```
Comment (kingjr) [inline neuralset-repo/neuralset/extractors/base.py]: I'm unclear about what s going on here

File: neuralset-repo/neuralset/extractors/base.py [hunk 10]
```diff
@@ -438,6 +437,120 @@ def _skip_new_event_types(key, val, prev):
...
+
+    @property
+    def torch_dtype_value(self) -> torch.dtype | tp.Literal["auto"] | None:
+        if self.torch_dtype is None or self.torch_dtype == "auto":
+            return self.torch_dtype
+        return getattr(torch, self.torch_dtype)
```
Comment (kingjr) [inline neuralset-repo/neuralset/extractors/base.py]: shall we set this at model_post_init and store as private attr?

File: neuralset-repo/neuralset/extractors/base.py [hunk 11]
```diff
@@ -438,6 +437,120 @@ def _skip_new_event_types(key, val, prev):
...
+    @property
+    def model_kwargs(self) -> dict[str, tp.Any]:
+        kwargs = self.config_kwargs.copy()
+        if self.device_map == "auto":
+            kwargs["device_map"] = self.device_map
+        if self.attn_implementation is not None:
+            kwargs["attn_implementation"] = self.attn_implementation
+        if self.torch_dtype_value is not None:
+            kwargs["torch_dtype"] = self.torch_dtype_value
+        return kwargs
```
Comment (kingjr) [inline neuralset-repo/neuralset/extractors/base.py]: I'm wondering whether we shouldn't instead use a private attr at the top of the class 
_CONFIG_VARS = ['revision', ..]
_MODEL_VARS = ['device_map', ...]

to avoid risking forgetting some elements in the future

File: neuralset-repo/neuralset/extractors/base.py [hunk 12]
```diff
@@ -479,63 +592,75 @@ class HuggingFaceMixin(base.BaseModel):
...
         if self.cache_n_layers is not None:
             excluded.extend(["layers", "layer_aggregation"])
         return excluded
 
+    @property
+    def device(self) -> str:
+        device_map = self.hf_config.device_map
+        if device_map == "auto":
+            return "cuda" if torch.cuda.is_available() else "cpu"
+        return device_map
```
Comment (kingjr) [inline neuralset-repo/neuralset/extractors/base.py]: at model_post_init ? not sure @jrapin ?

File: neuralset-repo/neuralset/extractors/image.py [hunk 13]
```diff
@@ -343,20 +261,35 @@ def model_post_init(self, log__):
...
+    def _full_predict(  # return the raw output, used in tests
+        self, images: torch.Tensor, text: str | list[str] = ""
+    ) -> tp.Any:
+        kwargs: dict[str, tp.Any] = dict(
+            images=[i.float() for i in images], return_tensors="pt"
+        )
+        if text:
+            kwargs["text"] = text
+        inputs = self.processor(**kwargs)
+        _fix_pixel_values(inputs)
```
Comment (kingjr) [inline neuralset-repo/neuralset/extractors/image.py:246]: this feels wrong. are you sure?

**Comment by kingjr** [comment]
You probably have the right design, but I'd love to have a live discussion with you and @jrapin  lucy to triple check

File: neuralset-repo/neuralset/events/transforms/test_transforms.py (hunk 1 above)
Comment (sdascoli) [reply neuralset-repo/neuralset/events/transforms/test_transforms.py]: Yeah it's a bit ugly I agree... I wanted model_config but protected by pydantic :(
If you have other ideas

File: neuralset-repo/neuralset/extractors/audio.py (hunk 2 above)
Comment (sdascoli) [reply neuralset-repo/neuralset/extractors/audio.py]: Agreed, let's simplify. This was because technically load_model loads the whole model while _load_cached_model only takes a part, like model.speech_encoder for audio, but who cares

File: neuralset-repo/neuralset/extractors/audio.py (hunk 3 above)
Comment (sdascoli) [reply neuralset-repo/neuralset/extractors/audio.py:468]: model_cls_name is used to specify the HF class when AutoModel doesn't work

File: neuralset-repo/neuralset/extractors/audio.py (hunk 4 above)
Comment (sdascoli) [reply neuralset-repo/neuralset/extractors/audio.py:487]: Here's my rationale:

Why model_name not in hf_config:
- To keep previous API intact
- Because most of the time you would leave the default hf_config, and it's easier to write `HuggingFaceText(model_name="gpt2")` than `HuggingFaceText(hf_config=dict(model_name="gpt2"))`

Why hf_config not in root: 
- Because I think it's quite self-contained (all the arguments which are passed to AutoModel.from_pretrained()
- Because it would incur many arguments at the root, most of these arguments are only used in rare situations like for Sophia's benchmark or for non-standard models, so nice to hide them in an "advanced" subfield

File: neuralset-repo/neuralset/extractors/base.py (hunk 5 above)
Comment (sdascoli) [reply neuralset-repo/neuralset/extractors/base.py:711]: I believe so yes

File: neuralset-repo/neuralset/extractors/base.py (hunk 6 above)
Comment (sdascoli) [reply neuralset-repo/neuralset/extractors/base.py]: Yes maybe that would be clearer you're right

File: neuralset-repo/neuralset/extractors/base.py (hunk 7 above)
Comment (sdascoli) [reply neuralset-repo/neuralset/extractors/base.py]: In principle yes: this is to choose sdpa, flash_attention etc which are functionally equivalent, not "linear" variants of attention

File: neuralset-repo/neuralset/extractors/base.py (hunk 8 above)
Comment (sdascoli) [reply neuralset-repo/neuralset/extractors/base.py]: We need to disambiguate config_kwargs and model_kwargs. Renaming to config_build_kwargs and model_build_kwargs for clarity

File: neuralset-repo/neuralset/extractors/base.py (hunk 10 above)
Comment (sdascoli) [reply neuralset-repo/neuralset/extractors/base.py]: Removing this method, only used once

File: neuralset-repo/neuralset/extractors/base.py (hunk 11 above)
Comment (sdascoli) [reply neuralset-repo/neuralset/extractors/base.py]: What would be the advantage? 
I'm not a fan of privatte attrs, and think it's better to remain stateless:
- They're ugly and not easy to understand for newcomers
- I have had issues in the past with private attrs, eg if you override model_post_init in a subclass and forget to call super().model_post_init(), or if you reinstantiate a class with model_dump

File: neuralset-repo/neuralset/extractors/base.py (hunk 12 above)
Comment (sdascoli) [reply neuralset-repo/neuralset/extractors/base.py]: Same as above, I prefer properties than private attrs for stuff which is instantaneous to compute

File: neuralset-repo/neuralset/extractors/image.py (hunk 13 above)
Comment (sdascoli) [reply neuralset-repo/neuralset/extractors/image.py:246]: This is copied from previous behaviour, I think it's to add caption

**Comment by sdascoli** [comment]
> You probably have the right design, but I'd love to have a live discussion with you and @jrapin lucy to triple check

Thanks for the quick and thorough feedback, just pushed some changes to address it!

File: neuralset-repo/neuralset/extractors/base.py (hunk 11 above)
Comment (kingjr) [reply neuralset-repo/neuralset/extractors/base.py]: really minor:
it would be a private attribute for the var names, not variables themselves. I think it's functionally equivalent, but allows it to be defined right below the class params. (i.e. if we had new config/model params, we'll more likely see that we need to update those fields and/or we can add a model_post_init to check that all variables are either in model or in config kwargs)

File: neuralset-repo/neuralset/extractors/base.py (hunk 12 above)
Comment (kingjr) [reply neuralset-repo/neuralset/extractors/base.py]: scratch that, if we submit from a non-cuda node, we don't want to freeze the attr

File: neuralset-repo/neuralset/extractors/image.py (hunk 13 above)
Comment (kingjr) [reply neuralset-repo/neuralset/extractors/image.py:246]: ok, let's deal with this later on. can you add a # FIXME

File: neuralset-repo/neuralset/extractors/base.py [hunk 14]
```diff
@@ -479,63 +582,72 @@ class HuggingFaceMixin(base.BaseModel):
...
-                self._REPOS.append(name)
-                self._REPOS.sort()
-                try:
-                    fp.write_text("\n".join(self._REPOS))
-                except Exception:
-                    pass  # nevermind if there is no write permission
-                return True
-            except Exception:
-                return False
+        self.hf_config.check_model_instantiates(self.model_name)
```
Comment (jrapin) [inline neuralset-repo/neuralset/extractors/base.py]: i'd expect this is way too heavy for the model post init, is it really instantiating a model?? if yes, do remove, this is way too huge a cost for initializing a config

File: neuralset-repo/neuralset/extractors/video.py [hunk 15]
```diff
@@ -92,7 +89,6 @@ class HuggingFaceVideo(extractor_base.BaseExtractor, extractor_base.HuggingFaceM
         "julius>=0.2.7",
     )
     model_name: str = "MCG-NJU/videomae-base"
-    pretrained: bool = True
```
Comment (jrapin) [inline neuralset-repo/neuralset/extractors/video.py:95]: why is this removed?

File: neuralset-repo/neuralset/extractors/video.py [hunk 16]
```diff
@@ -134,6 +130,15 @@ def _validate_model_name(cls, model_name: str) -> str:
 
     def model_post_init(self, log__: tp.Any) -> None:
         super().model_post_init(log__)
+        if (
+            "vjepa2" in self.model_name.lower()
+            and self.hf_config.processor_cls_name != "AutoVideoProcessor"
+        ):
+            msg = (
+                "V-JEPA2 models require "
+                "hf_config.processor_cls_name='AutoVideoProcessor'."
```
Comment (jrapin) [inline neuralset-repo/neuralset/extractors/video.py]: is it specific to video? + i'd expect this value should be the default?

File: neuralset-repo/neuralset/extractors/video.py (hunk 15 above)
Comment (sdascoli) [reply neuralset-repo/neuralset/extractors/video.py:95]: Moved to HuggingFaceMixin

File: neuralset-repo/neuralset/extractors/video.py (hunk 16 above)
Comment (sdascoli) [reply neuralset-repo/neuralset/extractors/video.py]: Looks like VideoMAE doesn't use it :/ But maybe I should put model_name="vjepa2" and processor_cls_name="AutoVideoProcessor" indeed

File: neuralset-repo/neuralset/extractors/base.py (hunk 14 above)
Comment (sdascoli) [reply neuralset-repo/neuralset/extractors/base.py]: Changed, thanks

**Review by jrapin — APPROVED** [review]

**Comment by kingjr** [comment]
🚀🧹🧽🧼 thanks @sdascoli !
